### ISSUE TITLE
Got NullPointerException when tried to read AVRO file generated by avro 1.7.4

### ISSUE THREAD

**Issue by java8964** [databricks/spark-avro#24]
I tried to test the spark-avro latest from the master with existing AVRO files stored in Hadoop 2.2.0 HDFS, generated by AVRO 1.7.4 version.

I installed the Spark 1.2.0, and spark-avro of latest from trunk, the jar file version is spark-avro_2.10-0.1.jar.

When I tried to load a small 2.3M avro file in spark, I got the errors.  Here are the steps:

[yzhang@hostname spark-1.2.0-bin-2.2.0]$ bin/spark-shell --jars ~/lib/spark-avro_2.10-0.1.jar
Spark assembly has been built with Hive, including Datanucleus jars on classpath
.........
Using Scala version 2.10.4 (IBM J9 VM, Java 1.7.0)
.......................

scala> import org.apache.spark.sql.SQLContext
import org.apache.spark.sql.SQLContext

scala> val sqlContext = new SQLContext(sc)
sqlContext: org.apache.spark.sql.SQLContext = org.apache.spark.sql.SQLContext@3790301b

scala> import com.databricks.spark.avro._
import com.databricks.spark.avro._

scala> val tags = sqlContext.avroFile("hdfs://namenode:9000/data/tags/")

tags: org.apache.spark.sql.SchemaRDD =
SchemaRDD[0] at RDD at SchemaRDD.scala:108
== Query Plan ==
== Physical Plan ==
PhysicalRDD [account_id#0L,delta#1,tags#2], MappedRDD[2] at map at AvroRelation.scala:63

scala> tags.count()

15/02/04 16:18:04 INFO executor.Executor: Fetching http://10.20.95.141:32451/jars/spark-avro_2.10-0.1.jar with timestamp 1423084457400
15/02/04 16:18:04 INFO util.Utils: Fetching http://10.20.95.141:32451/jars/spark-avro_2.10-0.1.jar to /tmp/fetchFileTemp7286568015846601068.tmp
15/02/04 16:18:05 INFO executor.Executor: Adding file:/tmp/spark-787b4ea0-4289-4a53-953e-e7b7c9a282bd/spark-avro_2.10-0.1.jar to class loader
15/02/04 16:18:05 INFO rdd.HadoopRDD: Input split: hdfs://namenode:9000/data/tags/20150202094448-part-r-00000.avro:0+1197027
15/02/04 16:18:05 INFO rdd.HadoopRDD: Input split: hdfs://p2-bibigin101:9000/data/tags/20150202094448-part-r-00000.avro:1197027+1197027
15/02/04 16:18:05 ERROR executor.Executor: Exception in task 1.0 in stage 0.0 (TID 1)
java.lang.NullPointerException
	at java.io.StringReader.<init>(StringReader.java:61)
	at org.apache.avro.Schema$Parser.parse(Schema.java:943)
	at org.apache.avro.Schema.parse(Schema.java:992)
	at org.apache.avro.mapred.AvroJob.getInputSchema(AvroJob.java:65)
	at org.apache.avro.mapred.AvroRecordReader.<init>(AvroRecordReader.java:43)
	at org.apache.avro.mapred.AvroInputFormat.getRecordReader(AvroInputFormat.java:52)
	at org.apache.spark.rdd.HadoopRDD$$anon$1.<init>(HadoopRDD.scala:233)
	at org.apache.spark.rdd.HadoopRDD.compute(HadoopRDD.scala:210)
	at org.apache.spark.rdd.HadoopRDD.compute(HadoopRDD.scala:99)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:263)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:230)
	at org.apache.spark.rdd.MappedRDD.compute(MappedRDD.scala:31)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:263)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:230)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:35)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:263)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:230)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:35)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:263)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:230)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:35)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:263)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:230)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:68)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:41)
	at org.apache.spark.scheduler.Task.run(Task.scala:56)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:196)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1176)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.lang.Thread.run(Thread.java:853)
15/02/04 16:18:05 ERROR executor.Executor: Exception in task 0.0 in stage 0.0 (TID 0)

It looks like the exception thrown from the Avro. I understand the the AVRO version comes with Spark-Avro is 1.7.7. Is this related to avro version mismatch, or some problem from the spark-avro?

Thanks

**Comment by mkanchwala** [comment]
can you please try this one as well, you can try it either by spark-shell / spark-submit: 

https://github.com/mkanchwala/spark-avro

**Comment by java8964** [comment]
Hi, I found out the root cause of this. It has nothing to do with SparkAvro, but really a spark build problem.

The SparkAvro jar doesn't contains any Avro classes. They are coming from the Spark runtime.
Since SparkAvro requires Spark 1.2, so I checkout release of Spark 1.2.0, and build it strictly follows the instructions. The Hadoop cluster I want to test is on HDFS 2.2.0, so I built Spark 1.2.0 with Hadoop 2.2.0. The problem exists in that Spark 1.2.0 depends on Avro 1.7.6, but Hadoop 2.2.0 depends on Avro 1.7.4. If I checked the stacktrace above, the org.apache.avro.* class match with Avro 1.7.6, but org.apache.avro.mapred.* class match with Avro 1.7.4. I believe in the Spark build processing, this 2 different runtime version of jars are merged into together, one from Spark 1.2.0 dependence, one from Hadoop 2.2.0 dependence.

What I did is manually upgrade all Avro classes in Spark spark-assembly-1.2.0-hadoop2.2.0.jar to Avro 1.7.6 version, then NullPointerException problem is gone.

Thanks
